### PR TITLE
Scanerror signals the consumer if we should continue scanning for next record & whether to checkpoint.

### PR DESCRIPTION
**Pull request by etsangsplk**
https://github.com/harlow/kinesis-consumer/pull/44
Basically address the the comment in the above PR

Scanerror signals the consumer if we should continue scanning for next record & whether to checkpoint.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,9 +38,13 @@ func main() {
 	}
 
 	// start
-	err = c.Scan(context.TODO(), func(r *consumer.Record) bool {
+	err = c.Scan(context.TODO(), func(r *consumer.Record) consumer.ScanError {
 		fmt.Println(string(r.Data))
-		return true // continue scanning
+        // continue canning
```
Comment (harlow) [inline README.md]: > continue scanning

Comment (etsangsplk) [reply README.md]: fixed.

File: README.md [hunk 2]
```diff
@@ -38,9 +38,13 @@ func main() {
 	}
 
 	// start
-	err = c.Scan(context.TODO(), func(r *consumer.Record) bool {
+	err = c.Scan(context.TODO(), func(r *consumer.Record) consumer.ScanError {
 		fmt.Println(string(r.Data))
-		return true // continue scanning
+        // continue scanning
```
Comment (harlow) [inline README.md]: wondering if we could use `return nil` for the success case. so the function signature is 

```
c.Scan(context.TODO(), func(r *consumer.Record) error {
```

then in the client loop we'd need to try to cast the error as `consumer.ScanError` if an error is returned

Comment (etsangsplk) [reply README.md]: This is to address  https://github.com/harlow/kinesis-consumer/issues/43. To finish off the otehr PR that is kind of hanging. 
It looks like it mentions about supporting  when a consumer will want to not checkpoint, but continue scanning the shard, and when a consumer also want to checkpoint also .. Unsure if returning 1 state will suffice.  Also Scan is a convenience wrapper around ScanShard.
What do u think?

File: consumer.go [hunk 3]
```diff
@@ -156,31 +164,34 @@ func (c *Consumer) Scan(ctx context.Context, fn func(*Record) bool) error {
...
 	// loop records
 	for r := range recc {
-		if ok := fn(r); !ok {
+		scanError := fn(r)
+		if scanError.StopScan {
 			break
 		}
+		if scanError.Error != nil {
+			c.logger.Println(fmt.Errorf("record processing error: %v", scanError.Error))
+		}
```
Comment (harlow) [inline consumer.go]: can remove the indentation of skip checkpoint

```go
for r := range recc {
    scanError := fn(r)
    
    if scanError.StopScan {
      break
    }

    if scanError.SkipCheckpoint {
      continue
    }

    c.counter.Add("records", 1)

    err := c.checkpoint.Set(c.streamName, shardID, *r.SequenceNumber)
    if err != nil {
      return fmt.Errorf("set checkpoint error: %v", err)
    }
}
```

also feels like there might be a logic issue between "stop" and "skip" i.e. maybe someone wants to stop, but also checkpoint?

File: README.md (hunk 2 above)
Comment (harlow) [reply README.md]: maybe thread regular `error` back up to the `err` case?

**Comment by harlow** [comment]
Hi @etsangsplk sorry for the delays on getting this PR merged. I just want to make sure we get this interface "right" before merging.

I was shuffling things around a little in this branch:
https://github.com/harlow/kinesis-consumer/compare/master...hw-scanerror?expand=1

Thoughts?

The main issue for me is I haven't had the need for this code, so I'm a little apprehensive about breaking the current method signature.

**Comment by etsangsplk** [comment]
@harlow 
Thanks.  I know there are multiple teams on my side have use cases for best effort processing. I am hoping to avoid forking if there is a common ground. But this should be the last change on scan and checjpoit

**Pull request merged**